pieces/shared/menu-button/menu.tsx

```diff
@@ -0,0 +1,147 @@
+import {
+  FocusEventHandler,
+  forwardRef,
+  KeyboardEventHandler,
+  MouseEventHandler,
+  ReactNode,
+  useCallback,
+  useImperativeHandle,
+  useRef,
+} from 'react';
+
+import { css } from '@emotion/react';
+
+import { token } from '@atlaskit/tokens';
+
+import { fallbackColor } from '../fallback';
+
+type MenuProps = {
+  children: ReactNode;
+  onKeyDown: KeyboardEventHandler;
+  onClose: (args: { shouldResetFocus: boolean }) => void;
+};
+
+export type MenuHandle = {
+  focusFirst(): void;
+  focusLast(): void;
+  focusNext(): void;
+  focusPrev(): void;
+};
+
+const menuStyles = css({
+  display: 'flex',
+  flexDirection: 'column',
+  position: 'absolute',
+  background: token('elevation.surface.overlay', fallbackColor),
+  boxShadow: token('elevation.shadow.overlay', fallbackColor),
+  minWidth: 128,
+  minHeight: 48,
+  width: 'max-content',
+  zIndex: 1,
+  top: 'calc(100% + 8px)',
+  borderRadius: 3,
+  right: 0,
+  listStyleType: 'none',
+  padding: 0,
+  margin: 0,
+});
+
+const stopPropagation: MouseEventHandler = (event) => {
+  event.stopPropagation();
+};
+
+const Menu = forwardRef<MenuHandle, MenuProps>(function Menu(
+  { children, onKeyDown: onKeydownProp, onClose },
+  handleRef,
+) {
+  const internalRef = useRef<HTMLUListElement>(null);
+
+  useImperativeHandle(
```

Comment:
I generally find imperative handles to be a bit mind bendy. Can you please switch this just to be helper functions?

```ts
function isHTMLElement(el: Node | null): el is HTMLElement {
  return el instanceof HTMLElement;
}

function focusFirstChild(el: Element | null) {
  if (!el) {
    return;
  }

  if (isHTMLElement(el.firstChild)) {
    el.firstChild.focus();
  }
}

function focusLastChild(el: Element | null) {
  if (!el) {
    return;
  }

  if (isHTMLElement(el.lastElementChild)) {
    el.lastElementChild.focus();
  }
}

function focusNextChild(el: Element | null) {
  if (!el) {
    return;
  }
  const focused = el.querySelector(":focus");
  // error case
  if (!focused) {
    focusFirstChild(el);
    return;
  }
  const next = el.nextElementSibling;
  if (!isHTMLElement(next)) {
    focusFirstChild(el);
    return;
  }
}

```